### PR TITLE
enh(sab): Add profileUrl to SAB

### LINKED ISSUE #38944: Write profile URL into system address book contact
### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are interested into the same feature.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments. 

**Is your feature request related to a problem? Please describe.**

Nextcloud users have a profile page.

image

**Describe the solution you'd like**

Write the profile URL to the SAB contact if the profile is enabled.
Don't write the URL otherwise.
Optional: write a type or similar attribute that allows us to distinguish *any* URLs from Nextcloud profile URLs for https://github.com/nextcloud/contacts/issues/3365

**Describe alternatives you've considered**

N/a

**Additional context**

* URL property spec: https://www.rfc-editor.org/rfc/rfc6350.html#section-6.7.8

### PR DESCRIPTION
**Pull request by JohannesGGE**
* Resolves: #38944 

## TODO

- [ ] ...

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [x] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [x] Screenshots before/after for front-end changes
- [x] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

File: apps/dav/lib/CardDAV/Converter.php [hunk 1]
```diff
@@ -88,6 +92,11 @@ public function createCardFromUser(IUser $user): ?VCard {
 				case IAccountManager::PROPERTY_WEBSITE:
 					$vCard->add(new Text($vCard, 'URL', $property->getValue(), ['X-NC-SCOPE' => $scope]));
 					break;
+				case IAccountManager::PROPERTY_PROFILE_ENABLED:
+					if ($property->getValue()) {
+						$vCard->add(new Text($vCard, 'X-SOCIALPROFILE', $this->config->getSystemValue('overwrite.cli.url', '').'/index.php/u/'.$user->getUID(), ['TYPE' => 'PROFILE', 'X-NC-SCOPE' => IAccountManager::SCOPE_PUBLISHED]));
```
Comment (tcitworld) [inline apps/dav/lib/CardDAV/Converter.php:97]: **suggestion:** 
```suggestion
						$vCard->add(new Text($vCard, 'X-SOCIALPROFILE', $this->config->getSystemValue('overwrite.cli.url', '').'/index.php/u/'.$user->getUID(), ['TYPE' => 'NEXTCLOUD', 'X-NC-SCOPE' => IAccountManager::SCOPE_PUBLISHED]));
```

Comment (tcitworld) [inline apps/dav/lib/CardDAV/Converter.php:97]: **nitpick:** Can you use the IURLGenerator instead?

Comment (JohannesGGE) [reply apps/dav/lib/CardDAV/Converter.php:97]: Yes. I already supposed that this was not the best way to get the url. :pray:

File: apps/dav/lib/CardDAV/Converter.php [hunk 2]
```diff
@@ -88,6 +92,11 @@ public function createCardFromUser(IUser $user): ?VCard {
 				case IAccountManager::PROPERTY_WEBSITE:
 					$vCard->add(new Text($vCard, 'URL', $property->getValue(), ['X-NC-SCOPE' => $scope]));
 					break;
+				case IAccountManager::PROPERTY_PROFILE_ENABLED:
+					if ($property->getValue()) {
+						$vCard->add(new Text($vCard, 'X-SOCIALPROFILE', $this->urlGenerator->getBaseUrl().'/index.php/u/'.$user->getUID(), ['TYPE' => 'NEXTCLOUD', 'X-NC-SCOPE' => IAccountManager::SCOPE_PUBLISHED]));
```
Comment (tcitworld) [inline apps/dav/lib/CardDAV/Converter.php:97]: **nitpick (blocking):** 
```suggestion
						$vCard->add(new Text($vCard, 'X-SOCIALPROFILE', $this->urlGenerator->linkToRouteAbsolute('core.ProfilePage.index', ['targetUserId' => $user->getUID()]), ['TYPE' => 'NEXTCLOUD', 'X-NC-SCOPE' => IAccountManager::SCOPE_PUBLISHED]));
```

**Review by ChristophWurst — APPROVED** [review]

**Comment by ChristophWurst** [comment]
You get bonus point for adjusting https://github.com/nextcloud/server/blob/master/apps/dav/tests/unit/CardDAV/ConverterTest.php ;)

The test can be run with ``NOCOVERAGE=1 ./autotest.sh sqlite apps/dav/tests/unit/CardDAV/ConverterTest.php``

**Review by miaulalala — APPROVED** [review]

**Comment by ChristophWurst** [comment]
The test fails and has to be adjusted

**Review by ChristophWurst — APPROVED** [review]

**Pull request merged**